rendermodules/rough_align/downsample_mask_handler.py

```diff
@@ -0,0 +1,192 @@
+from renderapi.errors import RenderError
+from rendermodules.rough_align.schemas import (
+        DownsampleMaskHandlerSchema)
+from rendermodules.module.render_module import (
+        RenderModule, RenderModuleException)
+import glob
+import os
+import pathlib
+import numpy as np
+import cv2
+from six.moves import urllib
+from shapely.geometry import Polygon, Point
+
+
+example = {
+        "render": {
+           "host": "em-131fs",
+           "port": 8987,
+           "owner": "TEM",
+           "project": "17797_1R",
+           "client_scripts": "/allen/aibs/pipeline/image_processing"
+                             "/volume_assembly/render-jars/production/scripts"
+        },
+        "mask_dir": "/allen/programs/celltypes/workgroups/em-connectomics"
+                    "/danielk/split_masks/section_masks",
+        "stack": "em_2d_montage_solved_py_0_01_mapped",
+        "collection": "chunk_rough_align_point_matches",
+        "zMask": [10395, 10398],
+        "zReset": None
+        }
+
+
+def points_in_mask(mask, pts):
+    # openCV >= 4.0:
+    # contours, _ = cv2.findContours(
+    #         mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+    # openCV < 4.0:
+    _, contours, _ = cv2.findContours(
+            mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
+    maskpoly = Polygon(contours[0].squeeze())
+    mask_list = [1.0 if maskpoly.contains(Point(pt)) else 0.0
+                 for pt in np.array(pts).transpose()]
+    return mask_list
+
+
+def filter_match(m, tspecs):
+    tgroups = np.array([t.layout.sectionId for t in tspecs])
+    for pq in ['p', 'q']:
+        tspec_ind = np.argwhere(tgroups == m[pq + 'GroupId']).flatten()
+        if tspec_ind.size > 0:
+            mask_path = urllib.parse.unquote(urllib.parse.urlparse(
+                tspecs[tspec_ind[0]].ip['0'].maskUrl).path)
+            mask = cv2.imread(mask_path, 0)
+            m['matches']['w'] = points_in_mask(mask, m['matches'][pq])
+    return m
+
+
+def reset_match(m):
+    m['matches']['w'] = [1.0] * len(m['matches']['w'])
+    return m
+
+
+class DownsampleMaskHandler(RenderModule):
+    default_schema = DownsampleMaskHandlerSchema
+
+    def get_tilespec(self, z):
+        try:
+            tspecs = renderapi.tilespec.get_tile_specs_from_z(
+                    self.args['stack'],
+                    z,
+                    render=self.render)
+        except RenderError:
+            return None
+
+        if len(tspecs) != 1:
+            raise RenderModuleException(
+                "Expected 1 tilespec for z=%d, found %d" % (z, len(tspecs)))
+
+        return tspecs[0]
+
+    def add_mask_to_tilespec(self, z):
+        tspec = self.get_tilespec(z)
+        if not tspec:
+            return tspec
+
+        if tspec.ip[0].maskUrl is not None:
+            raise RenderModuleException(
+                    "Tilespec already has a mask for z = %d" % (z))
+
+        fnames = glob.glob(os.path.join(
+            self.args['mask_dir'], '%d_*.png' % z))
```

Comment:
Does this module only accept "png" masks while the other one takes "png" or "tif"?

Reply:
mirrored that same `mask_exts` schema entry